File: branches/sandbox/darrth/darrt_actions/scripts/push_plate_test.py
import roslib; roslib.load_manifest('darrt_actions')
import rospy
import actionlib
import darrt_actions.msg
import arm_navigation_msgs.srv

class Table:
    def __init__(self, pose, x_min, x_max, y_min, y_max):
        self.pose = pose
        self.x_min = x_min
        self.x_max = x_max
        self.y_min = y_min
        self.y_max = y_max

def get_table():

    scene_client = rospy.ServiceProxy
    ("/environment_server/get_planning_scene", 
     arm_navigation_msgs.srv.GetPlanningScene);

    scene = scene_client.call();
    objects = scene.collision_objects;
    ctable = None
    for o in objects:
        if o.id == "table":
            ctable = o
            break
    if not ctable:
        rospy.logerr("Unable to find table in collision map.")
        return None

    if ctable.shapes.size() != 1 or\
            ctable.shapes[0].type != ctable.shapes[0].BOX:
        rospy.logerr("Table must be a box")
        return None
    tpose = geometry_msgs.msg.PoseStamped()
    tpose.header.frame_id = ctable.header.frame_id
    tpose.pose = ctable.poses[0]
    return Table(tpose, -1.0*ctable.shapes[0].dimensions[0]/2.0,
                 ctable.shapes[0].dimensions[0]/2.0,
                 -1.0*ctable.shapes[0].dimensions[1]/2.0,
                 ctable.shapes[0].dimensions[1]/2.0)

def get_goal_points():
    for i in range(100):
        #and this dies because ti's a point cloud2

def main():
    client = actionlib.SimpleActionClient("darrt_planning/push_plate_action",
                                          darrt_actions.msg.PushPlateAction)
    goal = darrt_actions.msg.PushPlateGoal()
    goal.plate_name = "plate"
    goal.surface_name = "table"
    client.send_goal_and_wait(goal)

if __name__ == '__main__':
    rospy.init_node('blah')
    main()

File: branches/sandbox/darrth/darrt_actions/src/push_plate_test.cpp
#include <actionlib/client/simple_action_client.h>
#include <arm_navigation_msgs/CollisionObject.h>
#include <arm_navigation_msgs/SetPlanningSceneDiff.h>

#include <pcl_ros/point_cloud.h>
#include <pcl/point_types.h>
#include <pcl/ros/conversions.h>

#include "sushi_manipulation_msgs/PushPlateAction.h"
#include <darrt/utils.hh>

namespace an = arm_navigation_msgs;

ros::Publisher objpub;

void create_plate(an::CollisionObject &object, double radius,
		  double height, geometry_msgs::PoseStamped pose) {
  object.id = "plate";
  object.operation.operation =
    an::CollisionObjectOperation::ADD;
  object.header.frame_id = pose.header.frame_id;
  object.header.stamp = ros::Time(0);
  an::Shape shape;
  shape.type = shape.CYLINDER;
  shape.dimensions.resize(2);
  shape.dimensions[0] = radius;
  shape.dimensions[1] = height;
  object.shapes.push_back(shape);
  object.poses.push_back(pose.pose);
  objpub.publish(object);
}

bool get_plate() {

  ros::NodeHandle n;
  ros::ServiceClient client =
    n.serviceClient<an::SetPlanningSceneDiff>
    ("/environment_server/set_planning_scene_diff");
  an::SetPlanningSceneDiff srv;
  if (!client.call(srv)) {
    ROS_ERROR("Unable to get planning scene");
    return false;
  }
  std::vector<an::CollisionObject> &objects =
    srv.response.planning_scene.collision_objects;
  int objind = -1;
  for (size_t i = 0; i < objects.size(); i++) {
    if (!objects[i].id.compare("plate")) {
      objind = i;
      break;
    }
    if (objects[i].id.compare("table") &&
	objects[i].shapes.size() == 1 &&
	objects[i].shapes[0].dimensions.size() == 3) {
      objind = i;
    }
  }
  std::string objectstr;
  for (size_t i = 0; i < objects.size(); i++) {
    objectstr += " (" + objects[i].id + ")";
  }

  if (objind < 0) {
    ROS_ERROR("Unable to find plate in planning scene.  Objects are: %s.", 
	      objectstr.c_str());
    return false;
  }

  ROS_INFO("Objects are: %s.  Plate is %s.", objectstr.c_str(), 
	   objects[objind].id.c_str());
  if (!objects[objind].id.compare("plate")) {
    //assume we've also corrected for a thicker table
    return true;
  }


  //create a plate from the pose and dimension of the bounding box
  geometry_msgs::PoseStamped pose;
  pose.header = objects[objind].header;
  pose.pose = objects[objind].poses[0];
  an::CollisionObject plate;
  create_plate(plate, objects[objind].shapes[0].dimensions[0]/2.0,
	       objects[objind].shapes[0].dimensions[2], pose);
  //remove the box
  objects[objind].operation.operation = objects[objind].operation.REMOVE;
  objpub.publish(objects[objind]);
  return true;
}



int main(int argc, char **argv) {
  ros::init(argc, argv, "push_plate_tester");
  ros::NodeHandle n;

  objpub = n.advertise<an::CollisionObject>("/collision_object", 1);
  actionlib::SimpleActionClient
    <sushi_manipulation_msgs::PushPlateAction>
    ppc("/darrt_planning/r_push_plate_action");
  ROS_INFO("Waiting for push plate action");
  ppc.waitForServer();
  ROS_INFO("Found action");
  if (!get_plate()) {
    ROS_ERROR("Unable to find plate in collision map");
    return false;
  }
  sushi_manipulation_msgs::PushPlateGoal goal;
  goal.plate_id = "plate";
  goal.surface_id = "table";
  ROS_INFO("Sending goal");
  ppc.sendGoalAndWait(goal);  
  ROS_INFO("Returned");
}
